File: tasks/translation/output/Qwen/QwQ-32B-AWQ/codenet/mad/Java/C/s685734377_comments_removed.c
#include <stdio.h>
#include <string.h>

int main() {
    int n;
    scanf("%d", &n);
    char a[n][n+1];
    for (int i = 0; i < n; i++) {
        scanf("%s", a[i]);
    }
    long long P = 37, Q = 47;
    long long powP[n+1], powQ[n+1];
    powP[0] = powQ[0] = 1;
    for (int i = 1; i <= n; i++) {
        powP[i] = powP[i-1] * P;
        powQ[i] = powQ[i-1] * Q;
    }
    long long hashRow[n][n], hashCol[n][n];
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            hashRow[i][j] = 0;
            hashCol[i][j] = 0;
        }
    }
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            for (int k = 翻译结果 0; k < n; k++) {
                int modP = (j - k + n) % n;
                hashRow[i][k] += (unsigned char)a[i][j] * powP[modP];
                int modC = (i - k + n) % n;
                hashCol[j][k] += (unsigned char)a[i][j] * powP[modC];
            }
        }
    }
    int ans = 0;
    for (int A = 0; A < n; A++) {
        for (int B = 0; B < n; B++) {
            int ok = 1;
            for (int i = 0; i < n; i++) {
                int row = (i + A) % n;
                int col = (i + B) % n;
                if (hashRow[row][B] != hashCol[col][A]) {
                    ok = 0;
                    break;
                }
            }
            ans += ok;
        }
    }
    printf("%d\n", ans);
} // End of Code

File: tasks/translation/output/Qwen/QwQ-32B-AWQ/codenet/mad/Java/Python/s685734377_comments_removed.py
import sys

class FastScanner:
    def __init__(self, stream):
        self.stream = stream
        self.buffer = []
        self.ptr = 0

    def next(self):
        while self.ptr >= len(self.buffer):
            line = self.stream.readline()
            if not line:
                return None
            self.buffer = line.split()
            self.ptr = 0
        res = self.buffer[self.ptr]
        self.ptr += 1
        return res

    def nextInt(self):
        return int(self.next())

class TaskB:
    def solve(self, testNumber, in_stream, out_stream):
        in_scanner = FastScanner(in_stream)
        n = in_scanner.nextInt()
        a = []
        for _ in range(n):
            line = in_scanner.next()
            a.append(list(line))
        P = 37
        Q = 47
        powP = [1] * (n + 1)
        powQ = [1] * (n + 1)
        for i in range(1, n + 1):
            powP[i] = powP[i - 1] * P
            powQ[i] = powQ[i - 1] * Q

        hashRow = [[0] * n for _ in range(n)]
        hashCol = [[0] * n for _ in range(n)]

        for i in range(n):
            for j in range(n):
                for k in range(n):
                    c = a[i][j]
                    val = ord(c)
                    exp_p_row = (j - k) % n
                    term_row = val * powP[exp_p_row]
                    hashRow[i][k] += term_row

                    exp_p_col = (i - k) % n
                    term_col = val * powP[exp_p_col]
                    hashCol[j][k] += term_col

        ans = 0
        for A in range(n):
            for B in range(n):
                ok = True
                for i in range(n):
                    ra = (i + A) % n
                    val_r = hashRow[ra][B]
                    rc = (i + B) % n
                    val_c = hashCol[rc][A]
                    if val_r != val_c:
                        ok = False
                        break
                if ok:
                    ans += 1
        out_stream.write(f"{ans}\n")

def main():
    in_stream = sys.stdin
    out_stream = sys.stdout
    solver = TaskB()
    solver.solve(1, in_stream, out_stream)
    out_stream.close()

if __name__ == "__main__":
    main()
# End of Code
